pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1277,74 +1279,87 @@ func getMatchingTopologyPairs(pod *v1.Pod, nodeInfoMap map[string]*schedulercach
 			catchError(fmt.Errorf("node not found"))
 			return
 		}
-		nodeTopologyMaps := newTopologyPairsMaps()
 		for _, existingPod := range nodeInfo.PodsWithAffinity() {
-			affinity := existingPod.Spec.Affinity
-			if affinity == nil {
-				continue
-			}
-			for _, term := range GetPodAntiAffinityTerms(affinity.PodAntiAffinity) {
-				namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(existingPod, &term)
-				selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
-				if err != nil {
-					catchError(err)
-					return
-				}
-				if priorityutil.PodMatchesTermsNamespaceAndSelector(pod, namespaces, selector) {
-					if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-						pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-						nodeTopologyMaps.addTopologyPair(pair, existingPod)
-					}
-				}
+			existingPodTopologyMaps, err := getMatchingTopologyPairsOfExistingPod(pod, existingPod, node)
+			if err != nil {
+				catchError(err)
+				return
 			}
-			if len(nodeTopologyMaps.podToTopologyPairs) > 0 {
-				appendTopologyPairsMaps(nodeTopologyMaps)
+
+			if len(existingPodTopologyMaps.podToTopologyPairs) > 0 {
+				appendTopologyPairsMaps(existingPodTopologyMaps)
 			}
 		}
 	}
 	workqueue.Parallelize(16, len(allNodeNames), processNode)
 	return topologyMaps, firstError
 }
 
+// getMatchingTopologyPairsOfExistingPod is a helper to calculate on each existing pod
+// of given node, whether (1) it has PodAntiAffinity and (2) AffinityTerms matches incoming pod
```

Comment:
I am not sure if I understand the logic in this function. Should we check whether `newPod` matches all the existing terms of the the `existingPod` and if it does, then we add the topology pairs of all the terms for this node to the return map?

Based on the above logic, I would write the function as:

```go
func getMatchingTopologyPairsOfExistingPod(newPod *v1.Pod, existingPod *v1.Pod, node *v1.Node) (*topologyPairsMaps, error) {
	affinity := existingPod.Spec.Affinity
	if affinity != nil && affinity.PodAntiAffinity != nil {
		terms := GetPodAntiAffinityTerms(affinity.PodAntiAffinity)
		for _, term := range terms {
			
			namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(existingPod, &term)
			selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
			if err != nil {
				return nil, err
			}
			// All terms must match for the affinity/anti-affinity rule to be considered a match.
			if !priorityutil.PodMatchesTermsNamespaceAndSelector(newPod, namespaces, selector) {
				return nil, nil
			}
		}
		// All terms have matched. Adding the topology pairs of this node...
		topologyMaps := newTopologyPairsMaps()
		for _, term := range terms {
			if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
				pair := topologyPair{key: term.TopologyKey, value: topologyValue}
				topologyMaps.addTopologyPair(pair, existingPod)
			}
		}
		return topologyMaps, nil
	}
	return nil, nil
}
```

Reply:
@bsalamat wouldn't it be better to use `targetPodMatchesAntiAffinityOfPod/targetPodMatchesAntiAffinityOfPod` directly? That would save time if the implementation changed (i.e the way terms match is changed) and avoid having duplicated code.

Reply:
> BTW: I'm a fan of do early return on negative path, then continue with happy path, which save s one loop of indenting, which is more readable.

@Huang-Wei +1. I like this style better too and agree that it is more readable.

> wouldn't it be better to use targetPodMatchesAntiAffinityOfPod/targetPodMatchesAntiAffinityOfPod directly? That would save time if the implementation changed (i.e the way terms match is changed) and avoid having duplicated code.

@ahmad-diaa Good point. It would be better to use targetPodMatchesAntiAffinityOfPod.

Reply:
@ahmad-diaa is new commit 8d9d0b48dce5c6b39980110d8bc1af9c0d530a6c what you meant?